core/src/entry_point/mod.rs

```diff
@@ -122,9 +122,7 @@ unsafe extern "C" fn rust_start(app_start: usize, stacktop: usize, _app_heap_bre
     // We get this from the environment to make it easy to set per compile.
     let app_heap_size: usize = read_APP_HEAP_SIZE();
 
-    // Make the heap start exactly at bss_end. The suggested _app_heap_break
-    // is almost always going to be too big and leads to us wasting memory.
-    let app_heap_start = bss_end;
+    let app_heap_start = app_heap_break;
```

Comment:
Unless I've completely lost track of how this works, this change is either going to leave a big empty space between `.bss` and the start of the heap or overlap the heap and `.bss`.

I don't see the relevance of the kernel-passed app_heap_break to userspace, it's pretty much just a fixed offset into the application's memory allocation.

Reply:
It shouldn't as we change app_heap_break in the assembly.

The problem is there is a bug in the assembly code.

Reply:
I miscounted arguments; the assembly no longer does what I thought it does.

Reply:
Actually, I think bss_end is still the right thing to put here. Not that it's *different* from app_heap_break, per se, but because we conceptually want to start the heap just after `.bss` ends.

Reply:
The assembly calculate `app_heap_break` to be after BSS and then passes that information to the kernel. That is the value we should be using as it matches what the kernel expects us to do.

Reply:
Ah, so you're saying this is the value we're using because it's where we told the kernel the heap would start using `memop()`? If yes, then I understand.

However, we need to rename the 3rd argument of `rust_start` to something other than `app_heap_break`. `app_heap_break` refers to the *end* of the heap, not the start, so its name is very misleading. At some point, the meaning of that argument was changed but its name was not updated to match.

Please rename `app_heap_break` to something like `heap_start`.